File: .lazy-harness/hooks/lifecycle/helpers/runtime_paths.py
#!/usr/bin/env python3
from __future__ import annotations

import hashlib
import json
import os
import subprocess
import time
from pathlib import Path
from typing import Any


def stable_hash(value: Any, length: int = 16) -> str:
    return hashlib.sha256(str(value or "").encode("utf-8", errors="replace")).hexdigest()[:length]


def payload_session_id(payload: dict[str, Any] | None = None) -> str:
    payload = payload or {}
    for key in ("session_id", "sessionId"):
        value = payload.get(key)
        if value:
            return str(value)
    for key in ("LAZY_SESSION_ID", "JCODE_SESSION_ID", "JCODE_SESSION"):
        value = os.environ.get(key)
        if value:
            return value
    return ""


def session_key(payload: dict[str, Any] | None = None, session_id: str | None = None) -> str:
    raw = session_id or payload_session_id(payload)
    if not str(raw or "").strip():
        return "default"
    return f"session-{stable_hash(raw, 20)}"


def git_output(root: Path, *args: str) -> str:
    try:
        return subprocess.check_output(["git", "-C", str(root), *args], text=True, stderr=subprocess.DEVNULL).strip()
    except Exception:
        return ""


def worktree_git_dir(root: Path) -> Path:
    out = git_output(root, "rev-parse", "--absolute-git-dir")
    if out:
        return Path(out).resolve()
    return root / ".lazy-harness" / ".gitless"


def git_common_dir(root: Path) -> Path:
    out = git_output(root, "rev-parse", "--git-common-dir")
    if not out:
        return worktree_git_dir(root)
    path = Path(out)
    return path.resolve() if path.is_absolute() else (root / path).resolve()


def runtime_root(root: Path, payload: dict[str, Any] | None = None, session_id: str | None = None) -> Path:
    explicit = os.environ.get("LAZY_RUNTIME_ROOT")
    if explicit:
        return Path(explicit).resolve()
    return worktree_git_dir(root) / "lazy-harness" / "runtime" / session_key(payload, session_id)


def shared_root(root: Path) -> Path:
    explicit = os.environ.get("LAZY_SHARED_ROOT")
    if explicit:
        return Path(explicit).resolve()
    return git_common_dir(root) / "lazy-harness" / "shared"


def runtime_state_path(root: Path, name: str, payload: dict[str, Any] | None = None, session_id: str | None = None) -> Path:
    return runtime_root(root, payload, session_id) / "state" / name


def runtime_log_path(root: Path, name: str, payload: dict[str, Any] | None = None, session_id: str | None = None) -> Path:
    return runtime_root(root, payload, session_id) / "logs" / name


def shared_path(root: Path, rel: str) -> Path:
    return shared_root(root) / rel


def append_jsonl_bounded(path: Path, row: dict[str, Any], keep: int = 200) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    existing: list[str] = []
    if path.exists():
        existing = [line for line in path.read_text(encoding="utf-8", errors="ignore").splitlines() if line.strip()][-(keep - 1):]
    existing.append(json.dumps(row, ensure_ascii=False, sort_keys=True))
    tmp = path.with_name(f"{path.name}.tmp-{os.getpid()}")
    tmp.write_text("\n".join(existing) + "\n", encoding="utf-8")
    tmp.replace(path)


def stable_json(value: Any) -> str:
    if isinstance(value, list):
        return "[" + ",".join(stable_json(item) for item in value) + "]"
    if isinstance(value, dict):
        return "{" + ",".join(
            json.dumps(key, ensure_ascii=False) + ":" + stable_json(val)
            for key, val in sorted(value.items())
            if val is not None
        ) + "}"
    return json.dumps(value, ensure_ascii=False)


def _acquire_lock(lock_dir: Path, timeout_seconds: float = 5.0) -> None:
    lock_dir.parent.mkdir(parents=True, exist_ok=True)
    started = time.time()
    while True:
        try:
            lock_dir.mkdir()
            (lock_dir / "owner.json").write_text(json.dumps({"pid": os.getpid(), "startedAt": time.time()}) + "\n", encoding="utf-8")
            return
        except FileExistsError:
            if time.time() - started > timeout_seconds:
                raise TimeoutError(f"Timed out waiting for lazy-harness lock: {lock_dir}")
            time.sleep(0.05)


def _release_lock(lock_dir: Path) -> None:
    try:
        for child in lock_dir.iterdir():
            child.unlink(missing_ok=True)
        lock_dir.rmdir()
    except Exception:
        pass


def append_jsonl_stable(root: Path, path: Path, row: dict[str, Any], id_key: str = "id") -> str:
    lock_dir = shared_root(root) / "locks" / f"jsonl-{stable_hash(path, 24)}.lockdir"
    _acquire_lock(lock_dir)
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        lines = [line for line in path.read_text(encoding="utf-8", errors="ignore").splitlines() if line.strip()] if path.exists() else []
        row_id = row.get(id_key)
        incoming = stable_json(row)
        for line in lines:
            try:
                existing = json.loads(line)
            except Exception:
                continue
            if stable_json(existing) == incoming:
                return "deduped-identical"
            if isinstance(row_id, str) and row_id:
                if isinstance(existing, dict) and existing.get(id_key) == row_id:
                    conflict = {
                        "id": f"conflict_{row_id}_{stable_hash(incoming, 12)}",
                        "event": "lazy-harness.jsonl-conflict",
                        "status": "conflict-recorded",
                        "detectedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                        "targetPath": str(path),
                        "idKey": id_key,
                        "conflictingId": row_id,
                        "existingHash": stable_hash(stable_json(existing)),
                        "incomingHash": stable_hash(incoming),
                        "incoming": row,
                    }
                    conflict_path = Path(str(path) + ".conflicts.jsonl")
                    with conflict_path.open("a", encoding="utf-8") as fh:
                        fh.write(json.dumps(conflict, ensure_ascii=False, sort_keys=True) + "\n")
                    return "conflict-recorded"
        with path.open("a", encoding="utf-8") as fh:
            fh.write(json.dumps(row, ensure_ascii=False, sort_keys=True) + "\n")
        return "appended"
    finally:
        _release_lock(lock_dir)


if __name__ == "__main__":
    import sys
    root = Path(os.environ.get("LAZY_HOST_ROOT") or os.getcwd()).resolve()
    payload: dict[str, Any] = {}
    if len(sys.argv) > 2 and sys.argv[2].strip():
        try:
            payload = json.loads(sys.argv[2])
        except Exception:
            payload = {}
    mode = sys.argv[1] if len(sys.argv) > 1 else "runtime-root"
    if mode == "runtime-root":
        print(runtime_root(root, payload))
    elif mode == "shared-root":
        print(shared_root(root))
    elif mode == "state-path":
        print(runtime_state_path(root, sys.argv[3] if len(sys.argv) > 3 else "state.json", payload))
    elif mode == "log-path":
        print(runtime_log_path(root, sys.argv[3] if len(sys.argv) > 3 else "log.jsonl", payload))

File: .lazy-harness/scripts/runtime-paths.ts
#!/usr/bin/env bun
import { createHash } from 'node:crypto'
import { existsSync, mkdirSync, readFileSync, writeFileSync, renameSync, rmSync } from 'node:fs'
import { dirname, join, resolve, isAbsolute } from 'node:path'
import { execFileSync } from 'node:child_process'

export function stableHash(value: unknown, length = 16): string {
  return createHash('sha256').update(String(value ?? ''), 'utf8').digest('hex').slice(0, length)
}

function gitOutput(root: string, args: string[]): string {
  try {
    return execFileSync('git', ['-C', root, ...args], { encoding: 'utf8', stdio: ['ignore', 'pipe', 'ignore'] }).trim()
  } catch {
    return ''
  }
}

export function hostRoot(explicit?: string): string {
  const candidate = explicit || process.env.LAZY_HOST_ROOT || process.cwd()
  return resolve(candidate)
}

export function sessionKey(sessionId?: string): string {
  const raw = sessionId || process.env.LAZY_SESSION_ID || ''
  const trimmed = String(raw || '').trim()
  if (!trimmed) return 'default'
  return `session-${stableHash(trimmed, 20)}`
}

export function worktreeGitDir(root = hostRoot()): string {
  const gitDir = gitOutput(root, ['rev-parse', '--absolute-git-dir'])
  if (gitDir) return gitDir
  return join(root, '.lazy-harness', '.gitless')
}

export function gitCommonDir(root = hostRoot()): string {
  const common = gitOutput(root, ['rev-parse', '--git-common-dir'])
  if (!common) return worktreeGitDir(root)
  return isAbsolute(common) ? common : resolve(root, common)
}

export function runtimeRoot(root = hostRoot(), sessionId?: string): string {
  const explicit = process.env.LAZY_RUNTIME_ROOT
  if (explicit) return resolve(explicit)
  return join(worktreeGitDir(root), 'lazy-harness', 'runtime', sessionKey(sessionId))
}

export function sharedRoot(root = hostRoot()): string {
  const explicit = process.env.LAZY_SHARED_ROOT
  if (explicit) return resolve(explicit)
  return join(gitCommonDir(root), 'lazy-harness', 'shared')
}

export function runtimeStatePath(name: string, root = hostRoot(), sessionId?: string): string {
  return join(runtimeRoot(root, sessionId), 'state', name)
}

export function runtimeLogPath(name: string, root = hostRoot(), sessionId?: string): string {
  return join(runtimeRoot(root, sessionId), 'logs', name)
}

export function sharedPath(rel: string, root = hostRoot()): string {
  return join(sharedRoot(root), rel)
}

export function ensureParent(filePath: string): void {
  mkdirSync(dirname(filePath), { recursive: true })
}

function sleep(ms: number): void {
  const end = Date.now() + ms
  while (Date.now() < end) {
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, Math.min(50, end - Date.now()))
  }
}

export function withLock<T>(lockName: string, fn: () => T, root = hostRoot(), timeoutMs = 5000): T {
  const lockDir = join(sharedRoot(root), 'locks', `${lockName.replace(/[^A-Za-z0-9_.-]/g, '_')}.lockdir`)
  mkdirSync(dirname(lockDir), { recursive: true })
  const started = Date.now()
  while (true) {
    try {
      mkdirSync(lockDir)
      writeFileSync(join(lockDir, 'owner.json'), JSON.stringify({ pid: process.pid, startedAt: new Date().toISOString() }) + '\n')
      break
    } catch {
      if (Date.now() - started > timeoutMs) {
        throw new Error(`Timed out waiting for lazy-harness lock: ${lockDir}`)
      }
      sleep(50)
    }
  }
  try {
    return fn()
  } finally {
    try { rmSync(lockDir, { recursive: true, force: true }) } catch { /* ignore */ }
  }
}

export type JsonlAppendStatus = 'appended' | 'deduped-identical' | 'conflict-recorded'

function stableJson(value: unknown): string {
  if (Array.isArray(value)) return `[${value.map(stableJson).join(',')}]`
  if (value && typeof value === 'object') {
    const entries = Object.entries(value as Record<string, unknown>)
      .filter(([, v]) => v !== undefined)
      .sort(([a], [b]) => a.localeCompare(b))
    return `{${entries.map(([k, v]) => `${JSON.stringify(k)}:${stableJson(v)}`).join(',')}}`
  }
  return JSON.stringify(value)
}

function atomicWrite(filePath: string, text: string): void {
  ensureParent(filePath)
  const tmp = `${filePath}.tmp-${process.pid}-${Date.now()}`
  writeFileSync(tmp, text, 'utf8')
  renameSync(tmp, filePath)
}

export function appendJsonlStable(filePath: string, row: Record<string, unknown>, idKey = 'id', root = hostRoot()): JsonlAppendStatus {
  const id = typeof row[idKey] === 'string' ? String(row[idKey]) : ''
  const incoming = stableJson(row)
  const lockName = `jsonl-${stableHash(filePath, 24)}`
  return withLock(lockName, () => {
    const lines = existsSync(filePath) ? readFileSync(filePath, 'utf8').split(/\r?\n/).filter((line) => line.trim()) : []
    for (const line of lines) {
      try {
        const existing = JSON.parse(line)
        if (stableJson(existing) === incoming) return 'deduped-identical'
        if (id) {
          if (existing && typeof existing === 'object' && existing[idKey] === id) {
            const conflict = {
              id: `conflict_${id}_${stableHash(stableJson(row), 12)}`,
              event: 'lazy-harness.jsonl-conflict',
              status: 'conflict-recorded',
              detectedAt: new Date().toISOString(),
              targetPath: filePath,
              idKey,
              conflictingId: id,
              existingHash: stableHash(stableJson(existing)),
              incomingHash: stableHash(stableJson(row)),
              incoming: row,
            }
            const conflictPath = `${filePath}.conflicts.jsonl`
            ensureParent(conflictPath)
            const current = existsSync(conflictPath) ? readFileSync(conflictPath, 'utf8') : ''
            writeFileSync(conflictPath, `${current}${JSON.stringify(conflict)}\n`, 'utf8')
            return 'conflict-recorded'
          }
        }
      } catch {
        // malformed historical lines do not block appending a new stable row
      }
    }
    lines.push(JSON.stringify(row))
    atomicWrite(filePath, `${lines.join('\n')}\n`)
    return 'appended'
  }, root)
}

if (import.meta.main) {
  const root = hostRoot()
  const sessionArg = process.argv.find((arg) => arg.startsWith('--session-id='))?.slice('--session-id='.length)
  const mode = process.argv[2] || 'print'
  if (mode === 'runtime-root') console.log(runtimeRoot(root, sessionArg))
  else if (mode === 'shared-root') console.log(sharedRoot(root))
  else if (mode === 'state-path') console.log(runtimeStatePath(process.argv[3] || 'state.json', root, sessionArg))
  else if (mode === 'log-path') console.log(runtimeLogPath(process.argv[3] || 'log.jsonl', root, sessionArg))
  else console.log(JSON.stringify({ hostRoot: root, runtimeRoot: runtimeRoot(root, sessionArg), sharedRoot: sharedRoot(root) }, null, 2))
}
